crates/messages/src/signatures.rs

```diff
@@ -5,79 +5,171 @@ use libp2p::{
+    QueryResult,
+};
+
+const SHA3_256_SIZE: usize = 32;
+const UUID_V4_SIZE: usize = 36;
 
-pub fn msg_hash<M: ProstMsg>(msg: &M) -> Vec<u8> {
-    let mut result = [0u8; 32];
+pub fn sha3_256(msg: &[u8]) -> [u8; 32] {
     let mut hasher = Sha3_256::default();
-    hasher.update(msg.encode_to_vec().as_slice());
-    Digest::finalize_into(hasher, result.as_mut_slice().into());
-    result.to_vec()
+    hasher.update(msg);
+    hasher.finalize().into()
 }
 
-fn verify_signature<T: SignedMessage>(peer_id: &PeerId, msg: &mut T) -> bool {
-    let sig = msg.detach_signature();
-    let encoded = msg.encode_to_vec();
-    let result = match PublicKey::try_decode_protobuf(&peer_id.to_bytes()[2..]) {
-        Ok(pubkey) => pubkey.verify(&encoded, &sig),
-        Err(_) => false,
-    };
-    msg.attach_signature(sig);
-    result
+fn sign(msg: &[u8], keypair: &Keypair) -> Vec<u8> {
+    keypair.sign(msg).expect("signing should succeed for Ed25519")
 }
 
-pub trait SignedMessage: ProstMsg + Sized {
-    fn detach_signature(&mut self) -> Vec<u8>;
-    fn attach_signature(&mut self, signature: Vec<u8>);
-
-    fn sign(&mut self, keypair: &Keypair) {
-        _ = self.detach_signature(); // To make signing idempotent
-        let bytes = self.encode_to_vec();
-        let signature = keypair.sign(&bytes).expect("infallible for Ed25519");
-        self.attach_signature(signature);
+fn verify_signature(peer_id: PeerId, msg: &[u8], signature: &[u8]) -> bool {
+    match PublicKey::try_decode_protobuf(&peer_id.to_bytes()[2..]) {
+        Ok(pubkey) => pubkey.verify(msg, signature),
+        Err(_) => false,
     }
+}
 
-    fn verify_signature(&mut self, peer_id: &PeerId) -> bool {
-        verify_signature(peer_id, self)
+impl Query {
+    pub fn sign(&mut self, keypair: &Keypair, worker_id: PeerId) -> Result<(), &'static str> {
+        let msg = self.signed_msg(worker_id).ok_or("couldn't sign invalid Query")?;
+        self.signature = sign(&msg, keypair);
+        Ok(())
     }
-}
 
-impl SignedMessage for Ping {
-    fn detach_signature(&mut self) -> Vec<u8> {
-        std::mem::take(&mut self.signature)
+    pub fn verify_signature(&self, signer_id: PeerId, worker_id: PeerId) -> bool {
+        if let Some(msg) = self.signed_msg(worker_id) {
+            verify_signature(signer_id, &msg, &self.signature)
+        } else {
+            false
+        }
     }
 
-    fn attach_signature(&mut self, signature: Vec<u8>) {
-        self.signature = signature;
+    fn signed_msg(&self, worker_id: PeerId) -> Option<Vec<u8>> {
+        // No two queries should have the same encoding. The easiest way to guarantee that
+        // is to use invertible encoding. That's why variable length fields are prefixed with their length.
+        if self.query_id.len() != UUID_V4_SIZE {
+            return None;
+        }
+        let worker_id_bytes = worker_id.to_bytes();
+        let mut msg = Vec::with_capacity(
```

Comment:
capacity will not match size if range is None, also worker_id_bytes's size not fixed.

Reply:
Extra capacity is not a problem, the size is correct anyway. It just avoida unnecessary reallocations